### PR TITLE
Reimplement View to support mutable Fork references [ECR-3525]

### PR DESCRIPTION
**Pull request by vitvakatu**
## Overview

New `View` implementation which works with mutable references and is more safe than the current.

## How to review

- All `*_index.rs` files contain a single change: no more need in dereferencing result of `View::get`
- `storage/db.rs` module was completely changed, review the file as whole, do not try to understand git diff.
- Take special interest in documentation (it is our internal docs, so we must understand it, but noone else must) and tests; there are a lot of corner-cases when dealing with such an unsafe API. It might not be possible (or useful) to check all of them, but if you have some ideas on additional test cases - you're welcome.
---
See: https://jira.bf.local/browse/ECR-3525

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

### CODE REVIEW

File: exonum-java-binding/core/rust/src/storage/db.rs [hunk 1]
```diff
@@ -28,104 +28,120 @@ pub(crate) type Value = Vec<u8>;
...
-/// way in Rust to make a `View` value not movable. Furthermore, it have to be moved from the stack
-/// to the heap in order to be converted into `Handle` for the java side. So a `Fork` value
-/// should be placed in the heap to prevent its movement after creating a reference to it.
-
-pub(crate) struct View {
-    /// Unsafe reference to the owned View (owned either by Rust or by Java)
-    reference: ViewRef,
-    /// Allows to drop owned view if needed
-    owned: Option<ViewOwned>,
+/// As `View` does not have a lifetime, nothing protects us from dereferencing a freed reference,
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/db.rs]: I don't get what you mean here by 'dereferencing a freed reference' and how it is connected with View instantiation.

File: exonum-java-binding/core/rust/src/storage/db.rs [hunk 2]
```diff
@@ -28,104 +28,120 @@ pub(crate) type Value = Vec<u8>;
...
+/// the extreme caution must be taken in places where `View` is constructed and send to Java.
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/db.rs]: sent to Java

File: exonum-java-binding/core/rust/src/storage/db.rs [hunk 3]
```diff
@@ -28,104 +28,120 @@ pub(crate) type Value = Vec<u8>;
...
+/// Java code must never store a `View` handle beyond the scope it was initially acquired
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/db.rs]: Not sure I understand what you mean here. Usually, Java releases a handle, and it must not be used subsequently. Is it anyhow different for this thing?

File: exonum-java-binding/core/rust/src/storage/db.rs [hunk 4]
```diff
@@ -28,104 +28,120 @@ pub(crate) type Value = Vec<u8>;
...
+    /// Created `View` must never outlive the reference it was created with,
+    /// or `SIGINT` will occur.
+    RefMutFork(&'static mut Fork),
+    /// Immutable Snapshot view, constructed from `&Snapshot`. There is no need in mutable
+    /// variant.
+    ///
+    /// Created `View` must never outlive the reference it was constructed with,
+    /// or `SIGINT` will occur.
+    RefSnapshot(&'static dyn Snapshot),
+    /// Covers both `Snapshot` and `Fork` cases. Rust uses move semantic and single-ownership
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/db.rs]: I'd clarify, what Owned means. If I understand correctly, the
client code owns a Snapshot/View object and then transfers the ownership of it to the View, which then owns it. I.e. 'owned' means different things in different contexts (= when used in factory method below and here in enum variant).

File: exonum-java-binding/core/rust/src/storage/db.rs [hunk 5]
```diff
@@ -141,95 +157,118 @@ mod tests {
...
+    fn get_mut_fork() {
+        let db = TemporaryDB::new();
+        let mut fork = db.fork();
+        let view = View::from_ref_mut_fork(&mut fork);
+        let mock_method = |_: &mut Fork| {};
+        match view {
+            View::RefMutFork(fork_ref) => {
+                mock_method(fork_ref);
+            }
+            _ => panic!(),
```
Comment (dmitry-timofeev) [inline exonum-java-binding/core/rust/src/storage/db.rs]: What is tested here and what the panics mean (also in other methods)?

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 1 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/db.rs]: The View does not have a lifetime, so we use `unsafe` to prolong the lifetime of the reference it was constructed with to `'static`. If this original reference is destroyed (leaves the scope), our prolonged reference stored inside View is no longer valid and will lead to SIGINT if we use it. So we must carefully review all the places where View is constructed from references to make sure View never outlives the original reference. Does it make sense now?

I agree, `freed` cannot be used in case of references here, I will rephrase.

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 3 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/db.rs]: Yep, no difference here. Java can't store handle to `View::Ref*` variants, because noone can verify that the reference will be still valid. Do you think I should remove this sentence as `NativeHandle` docs covers this enough? I just wasn't able to find any docs in native code explaining this mechanics.

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 4 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/db.rs]: You're right, but I can't see why the meaning in different here and there. 🤔 We have "owned" Snaphot/Fork -> we use `from_owned` method -> we have "owned" View, i.e. "View that owns Snapshot/Fork", which is the opposite of "View that borrows Snapshot/Fork".

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 5 above)
Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/db.rs]: It's just a pattern matching that has only one valid variant (and other variants must lead to test failure). I'd add panic message explaining the purpose.

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 1 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/db.rs]: Yes, thank you, that's clearer

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 3 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/db.rs]: > Java can't store handle to View::Ref* variants, because noone can verify that the reference will be still valid.

Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/db.rs]: What do you mean by Java not being able to store a handle to ref variants? How could it use them then? Or do you mean for longer than the method invocation?

Comment (vitvakatu) [reply exonum-java-binding/core/rust/src/storage/db.rs]: Yep, for longer than the method invocation (that was the purpose of "beyond the scope it was initially acquired") 🤔 Probably I should avoid such phrases, my fault.

File: exonum-java-binding/core/rust/src/storage/db.rs (hunk 4 above)
Comment (dmitry-timofeev) [reply exonum-java-binding/core/rust/src/storage/db.rs]: I mean that the subject which owns the snapshot/fork is different, but that it probably isn't that important.
